src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+making that connection).  The first module loaded from ``entry-point.coffee``
+with any notable side-effect ``smc-react.coffee``.  This initializes a single
+instance of a class called ``AppRedux`` which it exports to other modules with
+the variable name ``redux``.
+
+``AppRedux`` is the driver for SMC's own very-high-level wrapper around Redux.
+Explaining this is difficult unless you've read at least the introduction to
+Redux earlier in this document, if not read and understood the full
+documentation for Redux.  ``AppRedux`` maintains a sort of Redux meta-store.
+It contains only a single actual Redux store (as created with
+``redux.createStore``), but this is used to manage any number of sub-stores
+represented by key/value pairs at the top level of the main store's state.  The
+reason for this is that each page in SMC's UI might have its own state that is
+mostly independent of the state of other pages.  For example the "account" page
+may have state that is mostly independent of the "project" page's state, so the
+full state of the application looks something like::
+
+    {
+        "account": { <...account page state...> },
+        "projects": { <...projects page state...>},
+        ...
+    }
+
+There is even a sub-store called "page" for managing the overall current
+state of the CLI, such as what the currently active tab is.
+
+There are a few reasons for organizing things this way:
+
+* It keeps the application state fairly sanely organized, with sub-states
+  for each page, and easy routing of actions to the sub-states the affect.
+
+* However, since the entire state is stored in a single Redux store (as
+  opposed to, say, having separate stores for each page) it is also
+  possible to produce actions that affect multiple pages, or even other
+  parts of the application state that are not tied to a particular page or
+  view.  For example, the ``<Page>`` component, which we'll look at later,
+  connects to several different stores.
+
+* The ``AppRedux`` class makes it possible for each page/view to independently
+  and dynamically register a sub-store for itself.  The ``AppRedux`` instance
+  that is passed throughout the application then serves as a sane way to
+  manage all the known state stores.
+
+  In fact, much of ``AppRedux``'s API mimics the Redux library's own top-level
+  API.  So instead of calling |redux.createStore()|_ for each sub-store, one
+  actually calls ``AppRedux.createStore()`` (the latter has some important
+  differences from the former, however, which we'll come to later).  In fact,
+  since SMC names the ``AppRedux`` singleton ``redux``, one *does* in fact
+  literally call ``redux.createStore()``, but it's important to be clear that
+  here ``redux`` is an instance of ``AppRedux``, not the Redux library itself.
+
+The whole thing is fairly smart, and almost nothing about this framework is
+particular to SageMathCloud--it could (and probably should) be factored out
```

Comment:
In the future SMC is going to be a more generic application hosting framework...